### PR TITLE
feat: add Plan API to  typespec

### PR DESCRIPTION
**Pull request by chrisgacsal**
## Overview

Add initial version of Plan API to typespec.

### CODE REVIEW

File: api/spec/src/cloud/main.tsp [hunk 1]
```diff
@@ -140,3 +140,37 @@ namespace OpenMeterCloud.Entitlements {
 @route("/api/v1/customers")
 @tag("Customers")
 interface Customers extends OpenMeter.Customers {}
+
+namespace OpenMeterCloud.ProductCatalog {
```
Comment (GAlexIHU) [inline api/spec/src/cloud/main.tsp:157]: I suppose the contents of [this](https://github.com/openmeterio/openmeter/pull/1613/files) would also move under this namespace (eventually).
I'm ok with me sorting out the conflicts

File: api/spec/src/productcatalog/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# Plan Architecture
```
Comment (GAlexIHU) [inline api/spec/src/productcatalog/README.md:1]: I like the fact that we have extra docs 👍 

What would be a good way to (eventually) also incorporate these into the website's docs?

File: api/spec/src/productcatalog/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+| DELETE* | /api/v1/plans/`:planKey`/versions/`:planVersion`/phases/`:phaseKey` | Delete `PlanPhase` for `Plan`              |
+
+`PlanPhase` uniqueness constraints:
+* `planId` + `phaseKey`
+* `namespace` + `planKey` + `planVersion` + `phaseKey`
+
+## Open Questions
+| Q                                                                                            | A   |
+| -------------------------------------------------------------------------------------------- | --- |
+| Do we want to allow adding/deleting `PlanPhases` directly on `Plan` via phases API endpoint? | [ ] |
```
Comment (GAlexIHU) [inline api/spec/src/productcatalog/README.md:118]: nit: this probably belongs somewhere else

File: api/spec/src/productcatalog/plan.tsp [hunk 4]
```diff
@@ -42,13 +38,13 @@ model Plan extends UniqueResource {
    * The date and time when the plan becomes effective. When not specified, the plan is a draft.
    */
   @summary("Effective start date")
-  effectiveStartDate?: DateTime;
+  effectiveFrom?: DateTime;
```
Comment (GAlexIHU) [inline api/spec/src/productcatalog/plan.tsp:41]: So are we using effective or active?

File: api/spec/src/productcatalog/ratecards.tsp [hunk 5]
```diff
@@ -21,7 +16,16 @@ enum RateCardType {
  * Common fields of the rate card.
  */
 @friendlyName("RateCardBase")
-model RateCardBase extends Resource {
+model RateCardBase<T extends RateCardType> {
```
Comment (GAlexIHU) [inline api/spec/src/productcatalog/ratecards.tsp:19]: I made some changes to RateCards in https://github.com/openmeterio/openmeter/pull/1613/files#diff-b6b206b3b47ef8ef12e4ecd981d61c201a3de44ff2796ff866a0f67fb1a66e3cR33 could you take a look to see what you think?

I basically 1) did away with the Polymorphism 2) Extracted a part of it into Subscription (the part that actually gets created)

File: api/spec/src/productcatalog/routes.tsp [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+     * Filter by plan.key attribute
+     */
+    @query(#{ explode: true })
+    key?: string[],
+
+    /**
+     * Filter by plan.key and plan.version attributes
+     */
+    @query(#{ explode: true, style: "deepObject" })
+    keyVersion?: Record<integer>,
```
Comment (GAlexIHU) [inline api/spec/src/productcatalog/routes.tsp:41]: Are these all independent filters? 
e.g. `key=key-1&keyVersion[key-2]=v3` returns nothing, or would they have precedence?

**Review by GAlexIHU — DISMISSED** [review]
Nice!

File: api/spec/src/productcatalog/README.md (hunk 2 above)
Comment (tothandras) [reply api/spec/src/productcatalog/README.md:1]: Use `@opExample` in the API spec instead. We can use OpenAPI to generate docs.

File: api/spec/src/productcatalog/plan.tsp (hunk 4 above)
Comment (tothandras) [reply api/spec/src/productcatalog/plan.tsp:41]: `effective` ➕

File: api/spec/src/productcatalog/routes.tsp [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+  /**
+   * Create a new plan.
+   */
+  @post
+  @operationId("createPlan")
+  @summary("Create a plan")
+  create(@body request: TypeSpec.Rest.Resource.ResourceCreateModel<Plan>): {
+    @statusCode _: 201;
+    @bodyRoot body: Plan;
+  }  | OpenMeter.CommonErrors;
```
Comment (tothandras) [inline api/spec/src/productcatalog/routes.tsp:56]: `pnpm run format`

File: api/spec/src/productcatalog/routes.tsp (hunk 6 above)
Comment (chrisgacsal) [reply api/spec/src/productcatalog/routes.tsp:41]: So they are independent filters. Like `WHERE plan.id = 'x' OR plan.key = 'key-1' OR (plan.key='key-2' AND plan.version = 3)`

File: api/spec/src/cloud/main.tsp (hunk 1 above)
Comment (chrisgacsal) [reply api/spec/src/cloud/main.tsp:179]: Yup, moved it in this change.

File: api/spec/src/productcatalog/README.md (hunk 2 above)
Comment (chrisgacsal) [reply api/spec/src/productcatalog/README.md]: Removed teh README and will add `@opExample` latter.

File: api/spec/src/productcatalog/plan.tsp [hunk 8]
```diff
@@ -204,3 +206,20 @@ model PlanVariantOverridePreset {
   @summary("Discount")
   discount?: PlanDiscount;
 }
+
+/**
+ * Order by options for plans.
+ */
+@friendlyName("PlanOrderBy")
+enum PlanOrderBy {
```
Comment (tothandras) [inline api/spec/src/productcatalog/plan.tsp:214]: We should extend this eventually.

**Review by tothandras — APPROVED** [review]

**Review by tothandras — DISMISSED** [review]

**Review by tothandras — APPROVED** [review]

**Pull request merged**